src/core/ext/gcp/metadata_query.h

```diff
@@ -0,0 +1,72 @@
+//
+// Copyright 2023 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#ifndef GRPC_SRC_CORE_EXT_GCP_METADATA_QUERY_H
+#define GRPC_SRC_CORE_EXT_GCP_METADATA_QUERY_H
+
+#include <grpc/support/port_platform.h>
+
+#include <string>
+
+#include "absl/functional/any_invocable.h"
+#include "absl/status/status.h"
+#include "absl/status/statusor.h"
+
+#include "src/core/lib/gprpp/orphanable.h"
+#include "src/core/lib/gprpp/time.h"
+#include "src/core/lib/http/httpcli.h"
+#include "src/core/lib/http/parser.h"
+#include "src/core/lib/iomgr/closure.h"
+#include "src/core/lib/iomgr/polling_entity.h"
+
+namespace grpc_core {
+
+// Fetches the value of an attribute from the MetadataServer on a GCP
+// environment.
+class MetadataQuery : public InternallyRefCounted<MetadataQuery> {
+ public:
+  static const char kZoneAttribute[];
+  static const char kClusterNameAttribute[];
+  static const char kRegionAttribute[];
+  static const char kInstanceIdAttribute[];
+  static const char kIPv6Attribute[];
+
+  MetadataQuery(
+      std::string attribute, grpc_polling_entity* pollent,
+      absl::AnyInvocable<void(std::string /* attribute */,
+                              absl::StatusOr<std::string> /* result */)>
+          callback,
+      Duration timeout);
+
+  ~MetadataQuery() override;
+
+  void Orphan() override;
+
+ private:
+  static void OnDone(void* arg, absl::Status error);
```

Comment:
I think we should use `grpc_error_handle` instead of `absl::Status` as the type for this parameter, since this is a C-style closure.  That will be a signal to us that this is something that needs to be cleaned up later.

Reply:
because we want c-style closures to go away? Done